source/common/http/http1/codec_impl.cc

```diff
@@ -469,8 +469,42 @@ int ConnectionImpl::onHeadersCompleteBase() {
     protocol_ = Protocol::Http10;
   }
   if (Utility::isUpgrade(*current_header_map_)) {
-    ENVOY_CONN_LOG(trace, "codec entering upgrade mode.", connection_);
-    handling_upgrade_ = true;
+    // Ignore h2c upgrade requests until we support them.
+    // See https://github.com/envoyproxy/envoy/issues/7161 for details.
+    if (current_header_map_->Upgrade() &&
+        absl::EqualsIgnoreCase(current_header_map_->Upgrade()->value().getStringView(),
+                               Http::Headers::get().UpgradeValues.H2c)) {
+      ENVOY_CONN_LOG(trace, "removing unsupported h2c upgrade headers.", connection_);
+      current_header_map_->removeUpgrade();
+      if (current_header_map_->Connection()) {
+        auto values = Envoy::StringUtil::splitToken(
+            current_header_map_->Connection()->value().getStringView(), ",");
+        std::string new_values;
+        for (auto& v : values) {
+          auto value = StringUtil::trim(v);
+          auto lowercase_value = StringUtil::toLower(value);
+          if (lowercase_value == Http::Headers::get().ConnectionValues.Upgrade) {
+            continue;
+          }
+          if (lowercase_value == Http::Headers::get().ConnectionValues.Http2Settings) {
+            continue;
+          }
+          if (!new_values.empty()) {
+            new_values += ", ";
+          }
+          new_values.append(value.data(), value.size());
```

Comment:
I think this ends up N^2 on number of tokens unless we reserve space up front.  Given new_values is transient let's just reserve the space old_values takes up to avoid that.

Alternately I think you could std::remove elements you don't want and StringUtil::join the vector back together.

Reply:
No, because string allocation sizes increase by doubling the cost is O(N) in memory copied and O(NlgN) in number of allocations.  We could reserve the space but it isn't necessary.

We could remove the elements and join, but that would add an extra space at the front.  Envoy automatically adds a space after the header name, but splitTokens keeps the space on the first token.

Reply:
Sorry O(lnN) in allocations and O(N) in bytes copied.   In either case not DOS-able.

Reply:
Sorry if that was a bit abrupt, been working on too many DOS issues lately.